### PR TITLE
Replace NullableWalker.Result with TypeSymbolWithAnnotations

### PR DESCRIPTION
**Pull request by cston**
`Result` contained a type and a slot, and the slot can be calculated on demand.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -402,6 +402,7 @@ protected override int MakeSlot(BoundExpression node)
             switch (node.Kind)
             {
                 case BoundKind.ObjectCreationExpression:
+                case BoundKind.DynamicObjectCreationExpression:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >case [](start = 16, length = 4)

Can we cover this change with test?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -3323,26 +3319,20 @@ private TypeSymbolWithAnnotations GetTypeOrReturnTypeWithAdjustedNullableAnnotat
             return compilation.GetTypeOrReturnTypeWithAdjustedNullableAnnotations(symbol);
         }
 
-        private Result GetDeclaredLocalResult(LocalSymbol local)
+        private TypeSymbolWithAnnotations GetDeclaredLocalResult(LocalSymbol local)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >GetDeclaredLocalResult [](start = 42, length = 22)

Should the method be renamed? (ditto for following method)

**Comment by jcouv** [comment]
The change LGTM, but there are two failing tests in CI.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: This change is covered by `DynamicObjectCreation_03`. That test fails without this addition.

---
In reply to: [198627393](https://github.com/dotnet/roslyn/pull/28155#discussion_r198627393) [](ancestors = 198627393)

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please provide a second review.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -3323,26 +3319,20 @@ private TypeSymbolWithAnnotations GetTypeOrReturnTypeWithAdjustedNullableAnnotat
...
         {
-            var slot = GetOrCreateSlot(local);
             TypeSymbolWithAnnotations type;
-            if (!_variableTypes.TryGetValue(local, out type))
-            {
-                type = local.Type;
-            }
-            return Result.Create(type, slot);
+            return _variableTypes.TryGetValue(local, out type) ?
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >out type [](start = 53, length = 8)

Consider inlining declaration

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -1540,10 +1543,10 @@ public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperato
             }
 
             bool? resultIsNullable = getIsNullable(leftOperand, leftResult) == false ? false : getIsNullable(rightOperand, rightResult);
-            _result = TypeSymbolWithAnnotations.Create(resultType, resultIsNullable);
+            _resultType = TypeSymbolWithAnnotations.Create(resultType, resultIsNullable);
             return null;
 
-            bool? getIsNullable(BoundExpression e, Result r) => (r.Type is null) ? e.IsNullable() : r.Type.IsNullable;
+            bool? getIsNullable(BoundExpression e, TypeSymbolWithAnnotations t) => (t is null) ? e.IsNullable() : t.IsNullable;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Consider adding a newline between local function defs for readability, especially the one-line one.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -3072,17 +3070,16 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
...
+                TypeSymbolWithAnnotations rightType = ApplyConversion(right, right, conversion, leftType.TypeSymbol, rightResult, checkConversion: true, fromExplicitCast: false, out bool canConvertNestedNullability);
                 // Need to report all warnings that apply since the warnings can be suppressed individually.
                 ReportNullReferenceAssignmentIfNecessary(right, leftType, rightType, UseLegacyWarnings(left));
                 if (!canConvertNestedNullability)
                 {
-                    ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullabilityMismatchInAssignment, right.Syntax, GetTypeAsDiagnosticArgument(rightResult.Type?.TypeSymbol), leftType.TypeSymbol);
+                    ReportStaticNullCheckingDiagnostics(ErrorCode.WRN_NullabilityMismatchInAssignment, right.Syntax, GetTypeAsDiagnosticArgument(rightResult?.TypeSymbol), leftType.TypeSymbol);
                 }
-                TrackNullableStateForAssignment(right, leftType, leftResult.Slot, rightType, rightResult.Slot);
+                TrackNullableStateForAssignment(right, leftType, MakeSlot(left), rightType, MakeSlot(right));
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >MakeSlot(left) [](start = 65, length = 14)

Are there observable side-effects from moving the slot creation for `left` after visiting the right?

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Or is it just slots could be reordered and we don't care about that?

---
In reply to: [198932012](https://github.com/dotnet/roslyn/pull/28155#discussion_r198932012) [](ancestors = 198932012)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -3323,26 +3319,20 @@ private TypeSymbolWithAnnotations GetTypeOrReturnTypeWithAdjustedNullableAnnotat
...
+        private TypeSymbolWithAnnotations GetDeclaredParameterResult(ParameterSymbol parameter)
         {
-            var slot = GetOrCreateSlot(parameter);
             TypeSymbolWithAnnotations type;
-            if (!_variableTypes.TryGetValue(parameter, out type))
-            {
-                type = parameter.Type;
-            }
-            return Result.Create(type, slot);
+            return _variableTypes.TryGetValue(parameter, out type) ?
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >out type [](start = 57, length = 8)

Consider inlining declaration

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: The order that slots are created should not matter.

---
In reply to: [198932781](https://github.com/dotnet/roslyn/pull/28155#discussion_r198932781) [](ancestors = 198932781,198932012)

**Pull request merged**